File: backend/src/application/use-cases/active-cart/MergeCartUseCase.ts
import { v4 as uuidv4 } from "uuid";
import type { MergeCartUseCasePort } from "@application/ports/incoming/MergeCartUseCasePort";
import type { ActiveCartRepository } from "@application/ports/outgoing/ActiveCartRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type { MergeCartInput, MergeCartItemInput, MergeCartResult, MergeCartResultItem } from "./contracts";
import { ActiveCartMerger } from "@domain/services/ActiveCartMerger";
import type { ActiveCart, ActiveCartItem } from "@domain/entities/ActiveCart";

export class MergeCartUseCase implements MergeCartUseCasePort {
	constructor(
		private readonly activeCartRepository: ActiveCartRepository,
		private readonly logger: LoggerPort,
	) {}

	async execute(input: MergeCartInput): Promise<MergeCartResult> {
		this.logger.info(
			`[MergeCartUseCase] execute - userId: "${input.userId}", incomingItems: ${input.items.length}`,
		);

		const existingCart = await this.activeCartRepository.findByUserId(input.userId);
		const existingItems: readonly ActiveCartItem[] = existingCart?.items ?? [];
		const incomingItems = input.items.map(MergeCartUseCase.toActiveCartItem);
		const mergedItems = ActiveCartMerger.merge(existingItems, incomingItems);

		const cartToUpsert: ActiveCart = {
			id: existingCart?.id ?? uuidv4(),
			userId: input.userId,
			updatedAt: new Date(),
			items: mergedItems,
		};

		const savedCart = await this.activeCartRepository.upsert(cartToUpsert);

		this.logger.info(
			`[MergeCartUseCase] cart upserted - id: "${savedCart.id}", totalItems: ${savedCart.items.length}`,
		);

		return MergeCartUseCase.toResult(savedCart);
	}

	private static toActiveCartItem(item: MergeCartItemInput): ActiveCartItem {
		return {
			productId: item.productId,
			productName: item.productName,
			supermarket: item.supermarket,
			category: item.category,
			price: item.price,
			pricePerUnit: item.pricePerUnit,
			unit: item.unit,
			taxType: item.taxType,
			quantity: item.quantity,
			image: item.image,
			url: item.url,
		};
	}

	private static toResult(cart: ActiveCart): MergeCartResult {
		return {
			cartId: cart.id,
			userId: cart.userId,
			totalItems: cart.items.length,
			items: cart.items.map((item: ActiveCartItem) =>
				MergeCartUseCase.mapCartItemToResultItem(item),
			),
		};
	}

	private static mapCartItemToResultItem(item: ActiveCartItem): MergeCartResultItem {
		return {
			productId: item.productId,
			productName: item.productName,
			supermarket: item.supermarket,
			category: item.category,
			price: item.price,
			pricePerUnit: item.pricePerUnit,
			unit: item.unit,
			taxType: item.taxType,
			quantity: item.quantity,
			image: item.image,
			url: item.url,
		};
	}
}

File: backend-dotnet/src/api/Application/UseCases/ActiveCart/MergeCartUseCase.cs
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;
using SupermarketPlanner.Api.Domain.Entities;
using SupermarketPlanner.Api.Domain.Services;

namespace SupermarketPlanner.Api.Application.UseCases.ActiveCart;

public sealed class MergeCartUseCase(
    IActiveCartRepository activeCartRepository,
    ILogger<MergeCartUseCase> logger) : IMergeCartUseCase
{
    public async Task<Either<DomainError, MergeCartResult>> Invoke(
        MergeCartInput input,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "[MergeCartUseCase] Invoke - userId: \"{UserId}\", incomingItems: {Count}",
            input.UserId, input.Items.Count);

        var existingCart = await activeCartRepository.FindByUserId(input.UserId, cancellationToken);
        return await MergeAndUpsert(input, existingCart, cancellationToken);
    }

    private Task<Either<DomainError, MergeCartResult>> MergeAndUpsert(
        MergeCartInput input,
        Option<ActiveCartEntity> existingCart,
        CancellationToken cancellationToken)
    {
        var existingItems = existingCart.Match(onSome: cart => cart.Items, onNone: () => []);
        var incomingItems = input.Items.Select(ToActiveCartItem).ToList();
        var mergedItems = ActiveCartMerger.Merge(existingItems, incomingItems);

        var cartToUpsert = ActiveCartEntity.Create(
            id: existingCart.Match(onSome: cart => cart.Id, onNone: () => Guid.NewGuid().ToString()),
            userId: input.UserId,
            updatedAt: DateTimeOffset.UtcNow,
            items: mergedItems);

        return activeCartRepository.Upsert(cartToUpsert, cancellationToken)
            .MapAsync(savedCart =>
            {
                logger.LogInformation(
                    "[MergeCartUseCase] cart upserted - id: \"{CartId}\", totalItems: {Count}",
                    savedCart.Id, savedCart.Items.Count);
                return ToResult(savedCart);
            });
    }

    private static ActiveCartItem ToActiveCartItem(MergeCartItemInput item) =>
        new(item.ProductId, item.ProductName, item.Supermarket, item.Category,
            item.Price, item.PricePerUnit, item.Unit, item.TaxType, item.Quantity,
            item.Image, item.Url);

    private static MergeCartResult ToResult(ActiveCartEntity cartEntity) =>
        new(
            CartId: cartEntity.Id,
            UserId: cartEntity.UserId,
            TotalItems: cartEntity.Items.Count,
            Items: cartEntity.Items.Select(MapCartItemToResultItem).ToList());

    private static MergeCartResultItem MapCartItemToResultItem(ActiveCartItem item) =>
        new(item.ProductId, item.ProductName, item.Supermarket, item.Category,
            item.Price, item.PricePerUnit, item.Unit, item.TaxType, item.Quantity,
            item.Image, item.Url);
}
